### ISSUE TITLE
being able to specify how often a learning rule is applied

### ISSUE THREAD

**Issue by studywolf** [nengo/nengo#387]
Right now in my own learning code I'm doing a hack where I have a counter and apply the learning rule once every N timesteps instead of every timestep, just to speed up the simulation, it might be nice to have this option for learning rules in general.

**Comment by bmorcos** [comment]
Looking for some opinions on what is more intuitive. Let's say you want to apply a learning rule only **every 4 steps**, should you pass the parameter:
- As 4 (as in every 4 steps)
- As 1/4 (as in apply only 1/4 of the time)

Thoughts?

**Comment by jgosmann** [comment]
To me every `dt` seconds seems the most intuitive because everything else is usually specified in seconds and would make the value independent of the simulator `dt`. However, the simulator `dt` needs to be a multiple of the learning rule `dt` and that condition could be hard to ensure, also because of floating point errors (`sim_dt % lr_dt == 0` might be false even though analytically it should be true).

**Comment by tcstewar** [comment]
I definitely agree it should be specified in seconds, just like the Probe's ```sample_every``` argument.  And, indeed, that code also deals with the above problem of worrying about it being evenly divisible (or dealing with it if it isn't).  In that case, we just do:

```python
period = (1 if probe.sample_every is None else probe.sample_every / self.dt)
```

https://github.com/nengo/nengo/blob/master/nengo/simulator.py#L214

So that just rounds the number of steps to make it work.

**Comment by bmorcos** [comment]
Another discussion regarding the PES learning rule implementation in `nengo/builder/learning_rules.py`. Right now the PES rule does not implement it's own operator class like the other learning rules do (ex. `class SimBCM(Operator)`); instead PES uses general purpose operators. I was led to believe this has to do with the `nengo_OCL` implementation. I am looking at implementing this _learn sometimes_ feature in the operator by doing something like this:

```
    def make_step(self, signals, dt, rng):
        pre_filtered = signals[self.pre_filtered]
        post_filtered = signals[self.post_filtered]
        theta = signals[self.theta]
        delta = signals[self.delta]
        alpha = self.learning_rate * dt
        period = (1 if self.apply_every is None else apply_every / self.dt)

        def step_simbcm():
            delta[...] = np.outer(
                alpha * post_filtered * (post_filtered - theta),
                pre_filtered)

        def step_simbcm_sometimes():
            if n_steps % period < 1:
                delta[...] = np.outer(
                    alpha * post_filtered * (post_filtered - theta),
                    pre_filtered)

        if period > 1:
            n_steps = signals[self.n_steps]
            return step_simbcm_sometimes
        return step_simbcm
```
I think it makes more sense to create a specific `class SimPES(Operator)` instead of adding complexity to the general purpose operators (`ElementWiseInc` and `DotInc`). 

Would this affect other back ends? or is it just `nengo_OCL`? What do people think about this? Also please clarify if I've understood the implementation incorrectly :)

**Comment by jgosmann** [comment]
This could affect the optimizer as it implements optimizations for the a subset of the standard operators, but additional code is required for other operators.

On November 30, 2016 1:38:58 PM EST, Ben Morcos <notifications@github.com> wrote:
>Another discussion regarding the PES learning rule implementation in
>`nengo/builder/learning_rules.py`. Right now the PES rule does not
>implement it's own operator class like the other learning rules do (ex.
>`class SimBCM(Operator)`); instead PES uses general purpose operators.
>I was led to believe this has to do with the `nengo_OCL`
>implementation. I am looking at implementing this _learn sometimes_
>feature in the operator by doing something like this:
>
>```
>    def make_step(self, signals, dt, rng):
>        pre_filtered = signals[self.pre_filtered]
>        post_filtered = signals[self.post_filtered]
>        theta = signals[self.theta]
>        delta = signals[self.delta]
>        alpha = self.learning_rate * dt
>    period = (1 if self.apply_every is None else apply_every / self.dt)
>
>        def step_simbcm():
>            delta[...] = np.outer(
>                alpha * post_filtered * (post_filtered - theta),
>                pre_filtered)
>
>        def step_simbcm_sometimes():
>            if n_steps % period < 1:
>                delta[...] = np.outer(
>                    alpha * post_filtered * (post_filtered - theta),
>                    pre_filtered)
>
>        if period > 1:
>            n_steps = signals[self.n_steps]
>            return step_simbcm_sometimes
>        return step_simbcm
>```
>I think it makes more sense to create a specific `class
>SimPES(Operator)` instead of adding complexity to the general purpose
>operators (`ElementWiseInc` and `DotInc`). 
>
>Would this affect other back ends? or is it just `nengo_OCL`? What do
>people think about this? Also please clarify if I've understood the
>implementation incorrectly :)
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/nengo/nengo/issues/387#issuecomment-263956510

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by bmorcos** [comment]
Another point raised by @tbekolay, if we apply the learning rule only sometimes, ie.
```
period = (1 if self.apply_every is None else apply_every / self.dt)
def step_simvoja_sometimes():
    if n_steps % period < 1:
            delta[...] = alpha * learning_signal * (
                scale * np.outer(post_filtered, pre_decoded) -
                post_filtered[:, np.newaxis] * scaled_encoders)
    return step_simvoja_sometimes
```

Should we scale the `delta` signal by `apply_every / dt` to keep the learning at a similar rate despite not being applied every step? maybe @studywolf can weigh in about how you were using this (if you remember 😜 )

**Comment by jgosmann** [comment]
I didn't have the time to fully elaborate on my comment earlier: I don't know whether the optimizer actually speeds up learning rules based on general purpose operators. This would likely require some tests. Even if, I definitely see the argument that we might not want to make the general purpose operators more complex for this.

But then again, it seems that doing certain things only every n-th timestep might be something that we might want to do with more things (we already have this for probes). Maybe, it should be implemented in a more fundamental way on the operator level? (so that every operator would support it)

**Comment by bmorcos** [comment]
Update after quick a discussion in the dev meeting: I am planning to add scaling to the `delta` signal to make the learning rate more stable between normal learning and *learn sometimes*. I am also planning to implement the new `class SimPes(Operator)`. Maybe @hunse can comment on how much work it would be to implement this new class in `nengo_ocl`

**Comment by hunse** [comment]
> Maybe @hunse can comment on how much work it would be to implement this new class in `nengo_ocl`

Oh, not much. PES was implemented using basic operators not for nengo_ocl specifically, but just because we could, and there was no reason for writing a new operator. A side effect was that we got a PES implementation for free in nengo_ocl. But I've already implemented BCM and Oja there, so SimPES would be very straightforward.

**Comment by bmorcos** [comment]
I have code working that takes an argument `apply_every` in seconds, I then calculate the period (in terms of `dt` steps) at which to apply learning. If learning is not applied every time step (`period > 1`), then I am scaling the `delta` signal by `period` to keep learning at a similar rate. Here is an example of my PES implementation:
```
def make_step(self, signals, dt, rng):
        pre_filtered = signals[self.pre_filtered]
        delta = signals[self.delta]
        correction = signals[self.correction]
        learning_rate = self.learning_rate * dt
        n_neurons = self.n_neurons
        error = signals[self.error]
        period = (1 if self.apply_every is None else
                  self.apply_every / dt)

        if period > 1:
            n_steps = signals[self.n_steps]

            def step_simpesdecoders_sometimes():
                if n_steps % period < 1:
                    correction[...] = -learning_rate / n_neurons * error
                    delta[...] = np.outer(correction, pre_filtered)

                    # scale to compensate
                    delta[...] *= period
            return step_simpesdecoders_sometimes

        def step_simpesdecoders():
                    correction[...] = -learning_rate / n_neurons * error
                    delta[...] = np.outer(correction, pre_filtered)
        return step_simpesdecoders
```

This is working well for small values of `period` however once `period` starts to get larger, the `delta` signal becomes too large and the system has massive overshoot and oscillates around the input; no surprise here!

**Discussion time!**
- Do we definitely want to scale the `delta` signal?
- What would be an intelligent way to scale `delta`? 
    - Is an upper limit enough? 
    - do we want to do a more involved scaling algorithm?
- Maybe warn if `period` is large; or if we cap `period`, that learning will be slow?
- What does this look like for different learning rule types?

**Comment by tbekolay** [comment]
Out of curiosity, what values of `period` are you considering large?

**Comment by bmorcos** [comment]
This is the basic script I've been playing with in the GUI:
```
import nengo
from nengo.processes import WhiteSignal
import numpy as np

learning_rate = 1e-4
dt = 0.001  # not actually using this right now
apply_every = 30*dt

with nengo.Network() as model:
    stim = nengo.Node(WhiteSignal(60, high=5))
    pre = nengo.Ensemble(100, 1)
    normal = nengo.Ensemble(100, 1)
    sometimes = nengo.Ensemble(100, 1)
    error_normal = nengo.Ensemble(100, 1)
    error_sometimes = nengo.Ensemble(100, 1)

    nengo.Connection(stim, pre)
    nengo.Connection(normal, error_normal)
    nengo.Connection(pre, error_normal, transform=-1)
    nengo.Connection(sometimes, error_sometimes)
    nengo.Connection(pre, error_sometimes, transform=-1)
    conn_normal = nengo.Connection(
        pre, normal, function=lambda x: np.random.random(1),
        learning_rule_type=nengo.PES(learning_rate=learning_rate))
    conn_sometimes = nengo.Connection(
        pre, sometimes, function=lambda x: np.random.random(1),
        learning_rule_type=nengo.PES(learning_rate=learning_rate,
                                     apply_every=apply_every))
    nengo.Connection(error_normal, conn_normal.learning_rule)
    nengo.Connection(error_sometimes, conn_sometimes.learning_rule)
```

With that setup, once I get my `period` (`apply_every = period * dt`) to about 25-30 the performance starts to drop. This depends heavily on the learning rate though which is why I left it vague. If I bump up the learning rate, it follows that the value of `period` for which things become unstable is lower... Perhaps a better way to look at this is the *effective learning rate* as opposed to `period` proper; where *effective learning rate* would be something like `learning_rate * period`?

I haven't played around with other learning rules or values other than `dt=0.001`

**Comment by tbekolay** [comment]
Effective learning rate's an interesting thing to think about... There's also the mean magnitude of the error vector though, if you're using a big radius then you also have to scale down your learning rater, there are a lot of variables that go into these learning simulations! So I think I would be in favor of keeping it as is. Right now we raise a warning if the learning rate is unusually high, so perhaps we could add a warning if the period is unusually high also (say 200+ milliseconds). But they'd just be warnings; likely what is needed is a more in depth learning example showing the effects of tweaking all these parameters. If all that information can even be conveyed in an example! Regardless, I think they're things we don't have to worry about for your PR.
